### ISSUE TITLE
Validator as a Library: Suggested APIs

### ISSUE THREAD

**Issue by steveneliuk** [KhronosGroupArchives/webcl-validator#51]
During the Nov 14th, 2013 Telco it was agreed that some small changes to the APIs listed in public bugs 959 and 875 was needed.

```
Typedef void* ProgT
enum argInfo = { NAME, TYPE, ADDRESS_QUALIFIER, ACCESS_QUALIFIER };

ProgT ValidateProgram (const char * const * source, const char * const * activeWebCLExtList, const char * buildOptions);
bool isProgramValid (ProgT prog);
size_t getProgramValidatedSource (ProgT prog, size_t maxLen, char *buffer);
size_t getProgramValidatedSPIR (ProgT prog, size_t maSize, void *buffer);
size_t getKernelCount (ProgT prog);
size_t getKernelName(ProgT prog, size_t kernelID, size_t maxLen, char* buffer);
size_t getKernelArgCount(ProgT prog, size_t kernelID);
size_t getKernelArgInfo(ProgT prog, size_t kernelID, size_t argID, (enum) argInfo , size_t retValSize, void* retVal);
void destroyProgram(ProgT prog);
size_t getProgramLog (ProgT prog, size_t maxLen, char* buffer);
```

There was the suggestion to decompose getProgramLog() into getProgramErrors() and getProgramWarningsAndInfo(). Furthermore, if the effort required is minimal then to decompose getProgramWarningsAndInfo() into getProgramWarnings() and getProgramInfo().

**Comment by ollisal** [comment]
Hi,

I've been pondering this here at Vincit. My plan is to refactor the current monolithic webcl-validator executable to a library that exports a candidate version of this API and a thin wrapper executable offering the current validator functionality from the command line. This allows testing most of the library implementation with the current test suite of the webcl-validator CLI tool.

I'd like to get your comments on some design suggestions I have. Expect a proof of concept implementation along these lines in a week or two.

## 

First of all, to make it explicit, we should put everything in some prefix namespace, e.g. wclv (for WebCL Validator) to avoid any conflicts.

So we'd have wclvValidateProgram and so on. I'll omit the prefix for the rest of the discussion. 

I know this API has a pretty limited audience at this stage, but this future proofing measure is easy to make and likely helps in the long run.

## 

ValidateProgram() will be used in the browser implementations of WebCLProgram#build(). This JavaScript function is supposed to support passing an asynchronous callback as OpenCL compilation can take quite a bit of time. However, this is also true for the validator phase.

clBuildProgram has a callback argument, which can be used in an asynchronous implementation of WebCLProgram#build(). We should also add one to ValidateProgram, with the usual semantics i.e. can be called in a different thread, pass 0 to make it block, etc.

The rest of the API can be synchronous; it just reports an error/placeholder value if used when validation is still in progress. The intent is that one calls ValidateProgram() with a callback and only calls the rest of the API when the callback has been invoked.

Note that at least at this stage, although the API is async, the validation would not likely be done in parallel, but one program after another in a FIFO or LIFO fashion (unspecified which one). This is due to the Clang APIs used lacking thread safety. This can change later down the road, if this becomes a bottleneck e.g. with Web Workers. However we can still maintain the same API.

## 

Decomposing GetProgramLog() into GetProgramErrors() and the rest would make that API more semantic in terms of when there is a fatal error and when just some non-fatal warnings or debug info. Then again, existing similar Khronos APIs ranging from the GLSL glGetProgramInfoLog() to OpenCL clGetProgramBuildInfo() return just one log. However, these APIs report the status of building in a finer grained fashion.

Similarly, we could instead of

```
bool isProgramValid (ProgT prog);
```

have

```
enum ValidationStatus {
  // callback passed to ValidateProgram, validation not yet complete
  VALIDATION_IN_PROGRESS,
  // definitely illegal memory accesses
  VALIDATION_FATAL_ERROR,
  // there are some potentially unsafe operations, but might be OK
  VALIDATION_WARNINGS_EXIST,
  // all clear
  VALIDATION_SUCCESS
};

ValidationStatus GetProgramStatus(ProgT prog);
```

or somesuch. This enables e.g. rejecting the program altogether if there are errors, but accepting it and just logging the warnings for developers if those exist.

## 

The meaning of the size_t return value for getProgramValidatedSource, getKernelName etc is how many bytes have been written to buffer, right? For consistency with other OpenCL APIs, I'd make this a size_t \* output argument.

Instead, we should return a standard cl_int error from most of the functions.

## 

The rest of the API seems pretty much OK, though I'll likely make small adjustments wrt parameter types, naming etc.

**Comment by elhigu** [comment]
Are these still valid requirements for the API (from bugzilla 959 ticket)?
- extensions are given as a string array
- options are passed as a single string
- getValidationErrors(0, NULL) returns number of bytes required
- getValidationErrors(bufSize, buf) fills "buf" with the error string
- enum argInfo = { NAME, TYPE, ADDRESS_QUALIFIER, ACCESS_QUALIFIER }
- void *retVal may be int, char[], enum

Also I'm a bit confused about: 

```
    size_t getProgramValidatedSPIR (ProgT prog, size_t maSize, void *buffer);
```

Our API should not call OpenCL driver at any moment, so how should we be able to get SPIR or support SPIR? We do not even have complete LLVM linked in our library and we do not provide SPIR level instrumentation, but AST level.

**Comment by toaarnio** [comment]
Here's my take on the issues raised by @elhigu and @ollisal:
- The refactoring plan sounds good.
- I suggest the prefix "clv" instead of "wclv". The validator is not restricted to WebCL.
- Yes, we should definitely have a non-blocking form of clvValidateProgram.
- Yes, it would be good to have "machine-readable" error info in addition to the text log.
- I don't see what the SPIR getter is doing here, either. Am I missing something, @steveneliuk?
- The bugzilla ticket 959 requirements that you listed are still valid, AFAIK.
- I agree that the API should resemble OpenCL, e.g. return cl_int error codes, etc.

If there are no objections I'll draft an updated Validator API tomorrow with the above changes in place.

**Comment by toaarnio** [comment]
@ollisal from Vincit is currently refactoring the validator codebase into a proper library. He will draft an updated C API based on the above discussion.

**Comment by ollisal** [comment]
The current (partial) API implemented by PR #56:

``` c
// Run validation
clv_program clvValidate(
    const char *input_source,
    const char **active_extensions,
    const char **user_defines,
    void (CL_CALLBACK *pfn_notify)(clv_program program, void *user_data),
    void *notify_data,
    cl_int *errcode_ret);

typedef enum {
    /// Callback used, validation still running
    CLV_PROGRAM_VALIDATING,
    /// Illegal code encountered
    CLV_PROGRAM_ILLEGAL,
    /// Warnings from validation exist, but program can be run
    CLV_PROGRAM_ACCEPTED_WITH_WARNINGS,
    /// Program validated cleanly
    CLV_PROGRAM_ACCEPTED
} clv_program_status;

// Determine outcome of validation
clv_program_status clvGetProgramStatus(
    clv_program program);

// Get number of kernels found in program
cl_int clvGetProgramKernelCount(
    clv_program program);

// Get name of nth kernel in program
cl_int clvGetProgramKernelName(
    clv_program program,
    cl_uint n,
    size_t name_buf_size,
    char *name_buf,
    size_t *name_size_ret);

// Get number of arguments for the nth kernel in the program
cl_int clvGetKernelArgCount(
    clv_program program,
    cl_uint n);

// Get name of a kernel argument in the program
cl_int clvGetKernelArgName(
    clv_program program,
    cl_uint kernel,
    cl_uint arg,
    size_t name_buf_size,
    char *name_buf,
    size_t *name_size_ret);

// Get type of a kernel argument in the program
cl_int clvGetKernelArgType(
    clv_program program,
    cl_uint kernel,
    cl_uint arg,
    size_t type_buf_size,
    char *type_buf,
    size_t *type_size_ret);

// Determine if the given kernel argument is a pointer
cl_bool clvKernelArgIsPointer(
    clv_program program,
    cl_uint kernel,
    cl_uint arg);

// Return the address qualifier of the given pointer kernel argument
cl_kernel_arg_address_qualifier clvGetKernelArgAddressQual(
    clv_program program,
    cl_uint kernel,
    cl_uint arg);

// Determine if the given kernel argument is an image
cl_bool clvKernelArgIsImage(
    clv_program program,
    cl_uint kernel,
    cl_uint arg);

// Return the access qualifier of the given image kernel argument
cl_kernel_arg_access_qualifier clvGetKernelArgAccessQual(
    clv_program program,
    cl_uint kernel,
    cl_uint arg);

// Get validated source, ready to pass on to compiler
cl_int clvGetProgramValidatedSource(
    clv_program program,
    size_t source_buf_size,
    char *source_buf,
    size_t *source_size_ret);

// Release resources allocated by clvValidate()
void clvReleaseProgram(
    clv_program program);

```

Notes:
- extensions and user defines are passed as NULL-terminated C-string arrays, and these are the only options one can pass. These are the only clBuildProgram options that make sense in the validation context.
- OpenCL types, error codes and other conventions used as appropriate
- I decided to expand getKernelArgInfo to separate getters for each attribute; this is much cleaner to use and somewhat easier to implement with no real downside at just four attributes
- async callback to clvValidate ignored at the moment
- error reporting still missing (they go straight to the console); however it's possible to determine whether there were errors or was the validation successful

**Comment by elhigu** [comment]
Now error reporting is also implemented. For the first version all the necessary features seem to be implemented. I'll add separate ticket for writing API documentation, which refers to this issue.

**Comment by elhigu** [comment]
Backlink #83
